File: BaekjoonOnlineJudge/7576/7576.cpp
#include <iostream>
#include <queue>
#include <utility>

using namespace std;

int N, M;
int box[1001][1001];
bool visit[1001][1001];
queue<pair<int, int>> q;
int unRipedTomato;
int dx[4] = {0, 0, -1, 1};
int dy[4] = {-1, 1, 0, 0};

int BFS()
{
    int day = -1;

    while (q.size())
    {
        int qSize = q.size();
        day++;

        for (int oneDay = 0; oneDay < qSize; oneDay++)
        {
            int i = q.front().first;
            int j = q.front().second;
            q.pop();

            for (int way = 0; way < 4; way++)
            {
                int ii = i + dx[way];
                int jj = j + dy[way];

                if (ii < 0 || ii > N - 1 || jj < 0 || jj > M - 1)
                    continue;

                if (!visit[ii][jj] && box[ii][jj] != -1)
                {
                    visit[ii][jj] = true;
                    unRipedTomato--;
                    q.push({ii, jj});
                }
            }
        }
    }

    return day;
}

int main(void)
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int answer = 0;

    cin >> M >> N;

    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < M; j++)
        {
            cin >> box[i][j];

            if (box[i][j] == 0)
                unRipedTomato++;
            else if (box[i][j] == 1)
            {
                visit[i][j] = true;
                q.push({i, j});
            }
        }
    }

    if (!unRipedTomato)
        cout << 0;
    else
    {
        answer = BFS();

        if (unRipedTomato)
            cout << -1;
        else
            cout << answer;
    }

    return 0;
}

File: BaekjoonOnlineJudge/7576/7576.py
# 5:53 ~ 6:09 (16분)
from collections import deque
import sys

input = sys.stdin.readline


def bfs(dq):
    global answer, unriped_tomatos

    while dq:
        answer += 1

        for _ in range(len(dq)):
            i, j = dq.popleft()

            for way in range(4):
                ii = i + dx[way]
                jj = j + dy[way]

                if ii < 0 or ii > N - 1 or jj < 0 or jj > M - 1:
                    continue

                if box[ii][jj] == "0" and not visit[ii][jj]:
                    visit[ii][jj] = True
                    unriped_tomatos -= 1

                    dq.append((ii, jj))


answer = -1
M, N = map(int, input().split())

box = [list(input().split()) for _ in range(N)]
visit = [[False for _ in range(M)] for _ in range(N)]
unriped_tomatos = 0
dx = [0, 0, -1, 1]
dy = [1, -1, 0, 0]

dq = deque()

for i in range(N):
    for j in range(M):
        if box[i][j] == "0":
            unriped_tomatos += 1
        elif box[i][j] == "1":
            visit[i][j] = True
            dq.append((i, j))

bfs(dq)


if unriped_tomatos:
    print(-1)
else:
    print(answer)